url61db808cfe_ctriposs_bigcache_tgz-pJ8-com_ctriposs_bigcache_BigCacheStressTestJ8/src/com/ctriposs/bigcache/BigCache.java

```diff
@@ -0,0 +1,456 @@
+		this.storageManager.free();
+        /**
+         * we free storage first, so we can guarantee:
+         * 1. entries created/updated before "pointMap" clear process will not be seen. That's what free means.
+         * 2. entries created/updated after "pointMap" clear will be safe, as no free operation happens later
+         *
+         * There is only a small window of inconsistent state between the two free operation, and users should
+         * not see this if they behave right.
+         */
+        this.pointerMap.clear();
+        this.usedSize.set(0);
+	}
+
+	@Override
+	public double hitRatio() {
+		return 1.0 * hitCounter.get() / (hitCounter.get() + missCounter.get());
+	}
+
+	/**
+	 * Read Lock for key is locked.
+	 * 
+	 * @param key the key
+	 */
+	protected void readLock(K key) {
+		readWriteLock.readLock(Math.abs(key.hashCode()));
+	}
+
+	/**
+	 * Read Lock for key is unlocked.
+	 * 
+	 * @param key the key
+	 */
+	protected void readUnlock(K key) {
+		readWriteLock.readUnlock(Math.abs(key.hashCode()));
+	}
+
+	/**
+	 * Write Lock for key is locked..
+	 * 
+	 * @param key the key
+	 */
+	protected void writeLock(K key) {
+		readWriteLock.writeLock(Math.abs(key.hashCode()));
+	}
+
+	/**
+	 * Write Lock for key is unlocked.
+	 * 
+	 * @param key the key
+	 */
+	protected void writeUnlock(K key) {
+		readWriteLock.writeUnlock(Math.abs(key.hashCode()));
+	}
+
+    /**
+     * Get the internal lock.
+     * @param key
+     * @return
+     */
+    protected ReadWriteLock getLock(K key) {
+        return readWriteLock.getLock(Math.abs(key.hashCode()));
+    }
+
+	@Override
+	public void close() throws IOException {
+        this.clear();
+        this.ses.shutdownNow();
+		this.storageManager.close();
+	}
+	
+	public long count(){
+		return pointerMap.size();
+	}
+
+    /**
+     * Get the latest stats of the cache.
+     *
+     * @return all stats.
+     */
+    public BigCacheStats getStats() {
+        return new BigCacheStats(hitCounter.get(), missCounter.get(), getCounter.get(),
+                putCounter.get(), deleteCounter.get(), purgeCounter.get(), moveCounter.get(),
+                count(), storageManager.getUsed(), storageManager.getDirty(),
+                storageManager.getCapacity(), storageManager.getUsedBlockCount(), storageManager.getFreeBlockCount(),
+                storageManager.getTotalBlockCount());
+    }
+
+    abstract static class CacheDaemonWorker<K> implements Runnable {
+        private WeakReference<BigCache<K>> cacheHolder;
+        private ScheduledExecutorService ses;
+
+        CacheDaemonWorker(BigCache<K> cache) {
+            ses = cache.ses;
+            cacheHolder = new WeakReference<BigCache<K>>(cache);
+        }
+
+        @Override
+        public void run() {
+            BigCache cache = cacheHolder.get();
+            if (cache == null) {
+                // cache is recycled abnormally
+                if (ses != null) {
+                    ses.shutdownNow();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Call `shutdown()` to reject any incoming tasks and `awaitTermination()` to wait for existing tasks to complete or until a timeout is reached. To stop the execution and processing of tasks, if necessary, call `shutdownNow()`. Check the return value of `awaitTermination()` and handle when the method times out while waiting for other threads to stop executing. See the following diffs on GitHub that make this code change: [diff1](https://github.com/glowroot/glowroot/commit/81de5a3699c59e0cf98d485c380c1034e6cd28a5#diff-c608d2c40790b0e0e5228e6647f69729caa756bbe9127169139406314b80897eL143), [diff2](https://github.com/apache/directory-kerby/commit/3ed0e7cd3b8a0c3ce29293153e4e72f422f0637f#diff-eb42f072849f520563ec2a686789a836195f18075cc222ef355c6ea710b0a360L84)

Learn more: [ExecutorService in the Java Platform, Standard Edition 8 API Specification](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ExecutorService.html)